### ISSUE TITLE
Will this support svelte 5?

### ISSUE THREAD

**Issue by elron** [MacFJA/svelte-persistent-store#61]

**Comment by MacFJA** [comment]
This library is already supporting Svelte 5.

See #59 and version [`2.4.2`](https://github.com/MacFJA/svelte-persistent-store/releases/tag/2.4.2)

I use it on some SvelteKit + Svelte5 projects without any issues

**Comment by MacFJA** [comment]
**TL;DR**

It's possible, but not as neatly as with store. (worst DX than with store)

---

Runes are very complicated to customize.

Runes are compiler concept that are replaced at compilation time.  
They don't have any interface nor behavior contract nor type, so it's impossible to do Liskov substitution or high order function (wrapper) on them.  

The only possible way is with class, but that exclude all primitive value like: `let name = $state('John')` and it won't be a wrapper which is a bit limiting
And as runes are only available a compilation time, we can't have dynamic class (see below)

The best I can think of is this: 
```html
<script>
const name = persistState('John', createLocalStorage(), 'name')
</script>

<h1>Hello, {name.value}!</h1>
```
(`.value` is mandatory, no other solution)

_If you really want it, even with this limitation, it can be added_

---

**Limitation with dynamic class**

As runes are only available at compilation time and `$state` only available for property declaration and variable declaration in component root, I can't offer a solution like this

```html
<script>
const storage = createLocalStorage()
class Person extends PersistableState({ name: storage, age: storage}) {
  // ... your custom logic
}
const john = new Person({name: 'John', age: 33})
// On first launch, john will be `{name: 'John', age: 33}`
john.age = 34
// On later use, john will be loaded with age=34

const initial = $derived(john.name.substring(0, 1))
</script>

<h1>Hello, {john.name}!</h1>
<span>{initial}</span>
```
Which could be quite nice DX, but Svelte compiler don't allow it

---

It could exist a solution with a similar DX as store, but it will need to be a Vite plugin in order to "on the fly" replace code at compilation.
And this is far beyond the purpose of this project

---

_Also, it will not be (and will never be) a migration, but an addition, Store a still relevant in many cases_

**Comment by elron** [comment]
Thanks for thinking it thoroughly.
What is "DX" you mentioned a couple of times?

Anyway, I thought I'd show you what I do as an alternative in my project.


`store.svelte.ts` file (handles the states):
```ts
import { Persist } from "$lib/utils/Persist.svelte";

export const software = new Persist(
  "" as "fontcreator" | "glyphsapp" | "",
  "software"
);
```

`Persist.svelte.ts` file (handles all the logic, but only with local-storage at the moment - can be expanded for more use cases):

```ts
import { browser } from "$app/environment";

export class Persist<T> {
  #value: T = $state("" as T);
  #name: string | undefined = $state();

  constructor(defaultValue: T, name: string) {
    if (!browser) return;

    this.#name = name;
    const localValue = localStorage.getItem(name);
    if (localValue) {
      this.value = JSON.parse(localValue);
    } else {
      this.value = defaultValue;
      this.set(defaultValue);
    }
  }

  set(value: T) {
    // alert(this.#name);
    if (this.#name) {
      localStorage.setItem(this.#name, JSON.stringify(value));
    } else {
      alert(`this.#name is ${this.#name}`);
    }
  }

  set value(newValue: T) {
    if (this.#name && newValue) {
      this.set(newValue);
    }
    this.#value = newValue;
  }

  get value(): T {
    return this.#value;
  }
}
```

What do you think of this?

**Comment by MacFJA** [comment]
The more I think about it, the more I feel that a Vite plugin (or maybe a Rollup plugin) will be better.

We could achieve something like this:
```ts
let title = $persist('Unsaved document', 'document.title')
```
Then the plugin kick in and give this to the compiler
```ts
import { saveToStorage, loadFromStorage } from "..."
let title = $state(loadFromStorage(document.title) ?? 'Unsaved document')
$effect.root(() => {
  $effect(() => saveToStorage(title, 'document.title'))
})
```
So the variable can be used without the additional `.value`, it will be almost transparent as the diff between a `$state` and `$persist` is minimal:
```diff
- let title =   $state('Unsaved document')
+ let title = $persist('Unsaved document', 'document.title') 
```

**Comment by MacFJA** [comment]
> How would you implement TS type safety?

I didn't test this solution at all. But `$state` type declaration is just this:  https://github.com/sveltejs/svelte/blob/main/packages/svelte/src/ambient.d.ts#L23 
So it would be something like:
```ts
declare function $persist<T>(initial: T, key: string): T;
declare function $persist<T>(key: string): T | undefined;
```

**Comment by elron** [comment]
Sounds good!

**Comment by MacFJA** [comment]
@elron I just publish the first version of the lib:
- [Github](https://github.com/MacFJA/svelte-persistent-runes)
- [NPM](https://www.npmjs.com/package/@macfja/svelte-persistent-runes)

**Comment by elron** [comment]
Looks wonderful!! Good job!
